pgrx/src/callconv.rs

```diff
@@ -283,3 +283,182 @@ where
+        Self(fcinfo, std::marker::PhantomData)
+    }
+    /// Retrieve the arguments to this function call as a slice of [`pgrx_pg_sys::NullableDatum`]
+    #[inline]
+    pub fn raw_args<'a>(&'a self) -> &'fcx [pgrx_pg_sys::NullableDatum]
+    where
+        'a: 'fcx,
+    {
+        // Null pointer check already performed on immutable pointer
+        // at construction time.
+        unsafe {
+            let arg_len = (*self.0).nargs;
+            let args_ptr: *const pg_sys::NullableDatum = std::ptr::addr_of!((*self.0).args).cast();
+            // A valid FcInfoWrapper constructed from a valid FuntionCallInfo should always have
+            // at least nargs elements of NullableDatum.
+            std::slice::from_raw_parts(args_ptr, arg_len as _)
+        }
+    }
+
+    /// Modifies the contained `fcinfo` struct to flag its return value as null.
+    ///
+    /// # Examples
+    ///
+    /// ```rust,no_run
+    /// use pgrx::pg_return_null;
+    /// use pgrx::prelude::*;
+    /// fn foo(fcinfo: pg_sys::FunctionCallInfo) -> pg_sys::Datum {
+    ///     return unsafe { pg_return_null(fcinfo) };
+    /// }
+    /// ```
+    #[inline]
+    pub unsafe fn pg_return_null(&mut self) -> pg_sys::Datum {
+        let fcinfo = unsafe { self.0.as_mut() }.unwrap();
+        fcinfo.isnull = true;
+        pg_sys::Datum::from(0)
+    }
+
+    /// Get the collation the function call should use
+    #[inline]
+    pub unsafe fn pg_get_collation(&self) -> pg_sys::Oid {
+        let fcinfo = unsafe { self.0.as_mut() }.unwrap();
+        fcinfo.fncollation
+    }
+
+    /// Retrieve the type (as an Oid) of argument number `num`.
+    /// In other words, the type of `self.raw_args()[num]`
+    #[inline]
+    pub fn pg_getarg_type(&self, num: usize) -> pg_sys::Oid {
+        // Safety: User must supply a valid fcinfo to assume_valid() in order
+        // to construct a FcInfo. If that constraint is maintained, this should
+        // be safe.
+        unsafe {
+            pg_sys::get_fn_expr_argtype(self.0.as_ref().unwrap().flinfo, num as std::os::raw::c_int)
+        }
+    }
+
+    /// Retrieve the `.flinfo.fn_extra` pointer (as a PgBox'd type) from [`pg_sys::FunctionCallInfo`].
+    pub fn pg_func_extra<ReturnType, DefaultValue: FnOnce() -> ReturnType>(
+        &self,
+        default: DefaultValue,
+    ) -> PgBox<ReturnType> {
+        // Safety: User must supply a valid fcinfo to assume_valid() in order
+        // to construct a FcInfo. If that constraint is maintained, this should
+        // be safe.
+        unsafe {
+            let fcinfo = PgBox::from_pg(self.0);
+            let mut flinfo = PgBox::from_pg(fcinfo.flinfo);
+            if flinfo.fn_extra.is_null() {
+                flinfo.fn_extra = PgMemoryContexts::For(flinfo.fn_mcxt)
+                    .leak_and_drop_on_delete(default())
+                    as crate::void_mut_ptr;
+            }
+
+            PgBox::from_pg(flinfo.fn_extra as *mut ReturnType)
+        }
+    }
```

Comment:
For this, I basically consider the PgBox type soft-deprecated (there's not a good alternative atm or I would actually set the deprecation notice up), so I would rather we not use it here. It would still be safe, however, if it simply returned a raw pointer, as that would be unsafe to use.

Reply:
I've rewritten it to return a simpler pointer type, just a NonNull<T>.